proposals/0292-package-registry-service.md

```diff
@@ -38,77 +43,222 @@ it's not well-suited to this kind of workflow for the following reasons:
-| `GET`  | `/{package}`                                         | List package releases                            |
-| `GET`  | `/{package}/{version}`                               | Fetch metadata for a package release             |
-| `GET`  | `/{package}/{version}/Package.swift{?swift-version}` | Fetch manifest for a package release             |
-| `GET`  | `/{package}/{version}.zip`                           | Download source archive for a package release    |
-| `PUT`  | `/{package}/{version}{?commit,branch,tag,path,url}`  | Publish a package release or update its metadata |
+| Method | Path                                                      | Description                                     |
+| ------ | --------------------------------------------------------- | ----------------------------------------------- |
+| `GET`  | `/{scope}/{name}`                                         | List package releases                           |
+| `GET`  | `/{scope}/{name}/{version}`                               | Fetch metadata for a package release            |
+| `GET`  | `/{scope}/{name}/{version}/Package.swift{?swift-version}` | Fetch manifest for a package release            |
+| `GET`  | `/{scope}/{name}/{version}.zip`                           | Download source archive for a package release   |
+| `GET`  | `/identifiers{?url}`                                      | Lookup package identifiers registered for a URL |
 
 A formal specification for the package registry interface
 is provided alongside this proposal.
 In addition,
-an OpenAPI (v3) document and a reference implementation
+an OpenAPI (v3) document
+and a reference implementation written in Swift
 are provided for the convenience of developers interested
 in building their own package registry.
 
 ### Changes to Swift Package Manager
 
+#### Package identity
+
+Currently, the identity of a package is computed from
+the last path component of its effective URL
+(which can be changed with dependency mirroring).
+However, this approach can lead to a conflation of
+distinct packages with similar names
+and the duplication of the same package under different names.
+
+We propose using a scoped identifier
+in the form `scope.package-name`
+to identify package dependencies.
+
+A *scope* provides a namespace for related packages within a package registry.
+A package scope consists of
+alphanumeric characters and hyphens.
+Hyphens may not occur at the beginning or end,
+nor consecutively within a scope.
+The maximum length of a package name is 39 characters.
+A valid package scope matches the following regular expression pattern:
+
+```regexp
+\A[a-zA-Z\d](?:[a-zA-Z\d]|-(?=[a-zA-Z\d])){0,39}\z
+```
+
+A package's *name* is specified by the `name` provided in its manifest.
+
+The maximum length of a package name is 128 characters.
+A valid package name matches the following regular expression pattern:
+
+```regexp
+\A\p{XID_Start}\p{XID_Continue}{0,127}\z
+```
```

Comment:
cc @neonichu @abertelrud @yim-lee: are these general enough?

Reply:
I might be wrong, but I believe this is the same as what SwiftPM uses for package names today? 

I think scope and name should support the same characters but given the [thread on removing name from the manifest](https://github.com/apple/swift-evolution/pull/1301/files#r600880168) I am ok with leaving this as is.

Reply:
SwiftPM has no limit for package names